### PR TITLE
✨ Hollow Machines in CAPD

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**
This PR proposes an implementation of HollowMachines in Cluster API, based on kubemark/hollow kubelet.

The idea was originally proposed by @benmoss, but I took a different path in implementation so the delta diff is minimal and hopefully much easer to maintain…

**Which issue(s) this PR fixes:**
Fixes https://github.com/kubernetes-sigs/cluster-api/issues/4565

### CODE REVIEW

File: test/infrastructure/docker/api/v1alpha4/hollowmachine_types.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+package v1alpha4
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// HollowMachineFinalizer allows Reconcileto clean up resources associated with HollowMachine before
```
Comment (enxebre) [inline test/infrastructure/docker/api/v1alpha4/hollowmachine_types.go:25]: typo `Reconcileto`

File: test/infrastructure/docker/controllers/hollowmachine_controller.go [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+	log = log.WithValues("machine", machine.Name)
+
+	if util.IsControlPlaneMachine(machine) {
+		return ctrl.Result{}, errors.New("HollowMachine can't be used for control plane machines")
+	}
+
+	// Fetch the Cluster.
+	cluster, err := util.GetClusterFromMetadata(ctx, r.Client, machine.ObjectMeta)
+	if err != nil {
+		log.Info("HollowMachine owner HollowKubelet is missing cluster label or cluster does not exist")
```
Comment (enxebre) [inline test/infrastructure/docker/controllers/hollowmachine_controller.go:84]: should this log as error?

File: test/infrastructure/docker/controllers/hollowmachine_controller.go [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+	}
+
+	// Initialize the patch helper
+	patchHelper, err := patch.NewHelper(hollowMachine, r)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	// Always attempt to Patch the HollowMachine object and status after each reconciliation.
+	defer func() {
+		// TODO: investigating patching errors on delete
```
Comment (enxebre) [inline test/infrastructure/docker/controllers/hollowmachine_controller.go:107]: Personally I find the defer/patch/rerr pattern more difficult to follow than being explicit.
Why `if rerr == nil {`? `rerr` seems to be always nil.

**Comment by fabriziopandini** [comment]
/hold
while discussing if/how to consolidate with https://github.com/kubernetes-sigs/cluster-api-provider-kubemark

**Comment by elmiko** [comment]
i think this is cool, but i would like to maximize our efforts around kubemark. whether in capd or its own provider, i think it will be valuable.

for me, the main difference between this implementation and the standalone kubemark provider is that the latter presents an option for deploying onto non-capd clusters. otherwise it seems like _functionally_ they would be the same.

File: test/infrastructure/docker/api/v1alpha4/hollowmachine_types.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+	// Arguments to the kubemark entrypoint.
+	// Please note that --morph=kubelet and --name=<node name> will be always added to the list.
+	// +optional
+	Args []string `json:"args,omitempty" protobuf:"bytes,4,rep,name=args"`
+
+	// Compute Resources required by kubemark.
+	// if not specified cpu and memory requests, defaults of 40m and 10240Ki will be appied.
+	// +optional
+	Resources corev1.ResourceRequirements `json:"resources,omitempty"`
```
Comment (elmiko) [inline test/infrastructure/docker/api/v1alpha4/hollowmachine_types.go:78]: this is the resource requirements to run kubemark?

Comment (fabriziopandini) [reply test/infrastructure/docker/api/v1alpha4/hollowmachine_types.go:78]: Yes, those are resources for the container running the hollowkubelet

File: test/infrastructure/docker/controllers/hollowmachine_controller.go (hunk 3 above)
Comment (fabriziopandini) [reply test/infrastructure/docker/controllers/hollowmachine_controller.go:107]: defer is the standard approach in CAPI controllers.
`rerr` value depends on the return statements down in the same function

File: test/infrastructure/docker/controllers/hollowmachine_controller.go (hunk 2 above)
Comment (fabriziopandini) [reply test/infrastructure/docker/controllers/hollowmachine_controller.go:84]: we are alrading returning the error , which cause log.Error + exponential backoff

**Comment by sbueringer** [comment]
I took a rough look, looks very promising!

**Comment by elmiko** [comment]
this generally makes sense to me, and i agree with @sbueringer very promising!

i am currently trying to test this locally but hitting networking issues, i don't think it's related to the PR but i will report back if i cannot solve these issues.

**Comment by fabriziopandini** [comment]
/hold cancel
we agreed on keep developing this in CAPD to start exploring this space with SIG scalability

**Comment by elmiko** [comment]
been doing some more hacking on this, things are going well but i hit a small snag today when i updated my cluster-api code. i needed to rebase this patch, which required the following minor update:
```
diff --git a/test/infrastructure/docker/main.go b/test/infrastructure/docker/main.go
index afbb9c496..2a3b39043 100644
--- a/test/infrastructure/docker/main.go
+++ b/test/infrastructure/docker/main.go
@@ -33,7 +33,6 @@ import (
        "k8s.io/klog/v2"
        "k8s.io/klog/v2/klogr"
        clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
-       "sigs.k8s.io/cluster-api/controllers/remote"
        expv1 "sigs.k8s.io/cluster-api/exp/api/v1alpha4"
        "sigs.k8s.io/cluster-api/feature"
        infrav1old "sigs.k8s.io/cluster-api/test/infrastructure/docker/api/v1alpha3"
@@ -152,8 +151,8 @@ func setupReconcilers(ctx context.Context, mgr ctrl.Manager) {
        // Set up a ClusterCacheTracker to provide to controllers
        // requiring a connection to a remote cluster
        tracker, err := remote.NewClusterCacheTracker(
-               ctrl.Log.WithName("remote").WithName("ClusterCacheTracker"),
                mgr,
+               remote.ClusterCacheTrackerOptions{Log: ctrl.Log.WithName("remote").WithName("ClusterCacheTracker")},
        )
        if err != nil {
                setupLog.Error(err, "unable to create cluster cache tracker")
```

i'm hopeful to finish up my testing in the next few days, but i'm very optimistic about this work =)

**Comment by fabriziopandini** [comment]
Rebased!

File: test/infrastructure/docker/api/v1alpha4/hollowmachine_types.go [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+	// TODO: heapster, cluster-autoscaler
+	//  Note: Those components are not linked to a specific machines, but they are kubemark specific,
+	//  so there are two options: install using ClusterResourceSets, or make all the HollowMachines to reconcile them (create if missing).
+	//  Also: Are they required in the CAPI E2E scenario?
+
+	// TODO: dns; this component seems generic; what about using ClusterResourceSets?
+
+	// ExternalCluster define the name of the cluster where Pods hosting kubemark should be executed.
+	// If empty, the machine's cluster will be used.
+	ExternalCluster *string `json:"externalCluster"`
```
Comment (elmiko) [inline test/infrastructure/docker/api/v1alpha4/hollowmachine_types.go:45]: does this need an `omitempty` added to allow for the defaulting behavior?

**Review by elmiko — COMMENTED** [review]
just a question about one of the fields, also i think it would be really nice to have a readme with an example showing how to create a docker cluster that uses hollow nodes.

**Comment by elmiko** [comment]
i've been testing with some cluster templates that @fabriziopandini shared with me, and i'm hitting a snag that i'm having trouble debugging.

when i launch the external cluster (where the pods for the hollow cluster will run), i see it go into provisioning and it never becomes provisioned. i looked into the capd controller and saw a bunch of these errors at the end:

```
E0708 21:24:18.559716       1 controller.go:302] controller-runtime/manager/controller/dockermachine "msg"="Reconciler error" "error"="failed to pre-load images into the DockerMachine: command \"docker save -o /tmp/image-tar3286373947/image-0.tar gcr.io/k8s-staging-cluster-api/kubemark:v1.20.2\" failed with error: exit status 1" "name"="external-cluster-1-control-plane-tk7hv" "namespace"="default" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="DockerMachine" 
```

is this just a local configuration error?

do i need to pre-load some images into my local docker registry?

**Comment by sbueringer** [comment]
@elmiko I would have guessed you need the `gcr.io/k8s-staging-cluster-api/kubemark:v1.20.2` image in your local Docker engine. I got a very similar error yesterday when I forgot to build the manager images before running the tests, seems to say more or less "this local image is missing".

**Comment by elmiko** [comment]
ack, thanks @sbueringer 

i will try loading these into the local docker registry.

**Comment by elmiko** [comment]
had to build my own kubemark image, and load it locally. i couldn't access the one at gcr.io. so far, so good though, making progress.

**Comment by elmiko** [comment]
seem to be hitting a new error on the workload cluster, it gets stuck in provisioning and i see this in the capd-controller logs
```
E0709 18:27:13.812286       1 controller.go:302] controller-runtime/manager/controller/dockermachine "msg"="Reconciler error" "error"="failed to pre-load images into the DockerMachine: failed to load image: command \"docker exec --privileged -i test-1-control-plane-kxlr9 ctr --namespace=k8s.io images import -\" failed with error: exit status 1" "name"="test-1-control-plane-ddp97" "namespace"="default" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="DockerMachine" 
```

**Comment by fabriziopandini** [comment]
rebased
@elmiko I have also added examples of yaml for creating the external cluster and the test cluster with comments on the hollow specific bits. Hopefully this will help you in getting started in your environment

**Comment by elmiko** [comment]
been test driving the latest changes, and when i try to spawn a cluster from the `hollow-machines-external-cluster.yaml` i see this error after doing `kubectl apply`:

```Error from server (InternalError): error when creating "capd-hollow-cluster/hollow-machines-external-cluster.yaml": Internal error occurre
d: failed calling webhook "default.clusterresourceset.addons.cluster.x-k8s.io": the server could not find the requested resource```

i'm still digging into it, but just curious if anyone has seen this before?

**Comment by fabriziopandini** [comment]
> failed calling webhook "default.clusterresourceset.addons.cluster.x-k8s.io"

You need to turn on the CRS feature flag in your test env

**Comment by elmiko** [comment]
i was able to get past the CRS feature error and have hit a new problem. it appears that when i create the test cluster (the full kubemark cluster that will attach to the external cluster), i am seeing errors in the capd-controller logs that look like an open file error:

**capd controller logs**
```
I0720 20:25:29.046302       1 machine.go:476] controller-runtime/manager/controller/dockermachine "msg"="Got logs from the machine container" "name"="hollow-cluster-control-plane-t4xlt" "namespace"="default" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="DockerMachine" "output"=["INFO: ensuring we can execute mount/umount even with userns-remap","INFO: remounting /sys read-only","INFO: making mounts shared","INFO: detected cgroup v2","INFO: clearing and regenerating /etc/machine-id","Initializing machine ID from random ge
nerator.","INFO: faking /sys/class/dmi/id/product_name to be \"kind\"","INFO: faking /sys/class/dmi/id/product_uuid to be random","INFO: faking /sys/devices/virtual/dmi/id/product_uuid as well","INFO: setting iptables to detected mode: legacy","INFO: Detected IPv4 address: 172.18.0.8","INFO: Detected IPv6 address: ","systemd 246.6-1ubuntu1.1 running in system mode. (+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +ZSTD +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=hybrid)","Detected virtualization docker.","Detected architecture x86-64.","","Welcome to \u001b[1mUbuntu 20.10\u001b[0m!","","Set hostname to <hollow-cluster-control-plane-6vgd7>.","\u001b[0;1;31mFailed to create control group inotify object: Too many open files\u001b[0m","\u001b
[0;1;31mFailed to allocate manager object: Too many open files\u001b[0m","[\u001b[0;1;31m!!!!!!\u001b[0m] Failed to allocate manager object.","\u001b[0;1;31mExiting PID 1...\u001b[0m"]
E0720 20:25:29.046638       1 controller.go:304] controller-runtime/manager/controller/dockermachine "msg"="Reconciler error" "error"="failed to exec DockerMachine bootstrap: failed to run cloud config: command \"docker exec --privileged hollow-cluster-control-plane-6vgd7 mkdir -p /etc/kubernetes/pki\" failed with error: exit status 1" "name"="hollow-cluster-control-plane-t4xlt" "namespace"="default" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="DockerMachine" 
```

i thought maybe my system was setup with a low limit on these, but i double checked and it seems like it should be unlimited:
```
$ cat /proc/sys/fs/file-max 
9223372036854775807
$ ulimit
unlimited
```
and my docker service file seems to have unlimited as well:
```
LimitNOFILE=infinity
```

i'm not sure why it won't let that hostname file go through during bootstrap. i'm going to keep poking around to see if i can find more answers, but i figured this was worth raising to see if i am missing something.

**Comment by vincepri** [comment]
@fabriziopandini Should we try to finish this PR before next release?

**Comment by fabriziopandini** [comment]
@vincepri rebased
The PR is pretty stable and comments are about documentation/problems @elmiko local env. 
Personally I'm +1 to move this to the next stage and start trying to use it in some of our CI tests

**Comment by elmiko** [comment]
+1 to what @fabriziopandini is saying, i am very excited to see this get merged. i am struggling to get it working locally, so i would like to make sure it easy for others to reproduce this workflow. i think that's my main concern currently.

**Comment by fabriziopandini** [comment]
/close
As per office hours/slack discussions this will land in https://github.com/kubernetes-sigs/cluster-api-provider-kubemark

**Pull request closed without merging**